### ISSUE TITLE
Possibility for drag n drop samples?

### ISSUE THREAD

**Issue by alamminsalo** [dagargo/elektroid#11]
I tried to fiddle with drag-n-dropping samples from file browser, but I don't have any experience in gtk. This feature would make pretty good QOL improvement

**Comment by dagargo** [comment]
I'm currently working on this as I think I'd be a nice addition. However, I have some doubts I'd like to discuss.

I don't know exactly what to do when drag-and-dropping on the local panel and the source is not the remote panel. What would you expect to be done? Nothing? Files to be copied? Files to be moved?

Besides this, If you're gonna use mostly drag-and-drop, wouldn't it be nice to have a remote-only interface? Of course, this should be user selectable. We could hide the local panel and the audio controls, but not the task list.

Finally, I'd like to add that `elektroid -l dir` opens Elektroid and sets the local directory to `dir`. This is useful if you want to integrate Elektroid with your file browser.

**Comment by simonvanderveldt** [comment]
Not sure if this should be a different bug or not but I expected to be able to drag samples from the left pane to the right, same as is possible in Elektron Transfer. Would this be possible?
(Let me know if I should create a separate issue for this)

**Comment by dagargo** [comment]
I consider that this issue should cover drag and drop from left to right as well as from right to left. But I think it should also cover drag and drop from another applications.

Once this is done, copy and paste could be implemented easily, I'd say.

However, this is a lot of work. I tried to implement this and I think it's better to focus on new functionalities for now.

>  (Let me know if I should create a separate issue for this)

This issue is enough.

**Comment by dagargo** [comment]
I've just pushed some commits that provide basic in-application DND. Multiple selection is allowed but it is not possible to drag to a concrete directory on the destination at the moment.

We have two different scenarios for DND between elektroid and another application. I don't know if Elektron Transfer lets you do this.

* Dragging from elektroid is difficult. As the resources are not available as files, we need to provide a URL thru an embedded web server to let other applications access the resources. The problem with this approach is that if the transfers are time consuming, the applications receiving data from this web server might cancel the downloads due to timeouts. Also, remember that only one transfer is allowed at the same time.

* Dropping on elektroid is doable. But i find it weird to be able to drop but not to drag from.

Same scenarios apply to copy and paste.

BTW, this is a bit experimental so you might find some bugs.

Let we know what you think about this.

**Comment by dagargo** [comment]
Last commits fix some issues and complete in-application DND and add support for outside-application drag from local browser, drop on local browser and drop on remote browser.

Only outside-application drag from remote browser is pending.

**Comment by dagargo** [comment]
It's been a long time but DND implementation is completed.

Now, it's possible to drop on a concrete directory for local and device browser.

The only DND functionality that is not implemented is when dragging from the device browser and dropping into your file manager. It's too complicated and so limited by the DND system that makes not sense to implement it.

It will be very nice if you all could do some testing and share your thoughts on this. 

TIA.

**Comment by dagargo** [comment]
To make DND more user friendly, I've just pushed a commit that allows the user to go up in the directory tree while dragging.

You just need to drag the cursor over the up directory button and leave it there for 1 second. If you leave it there more time it will be going up in the directory hierarchy (one level up every second).

**Comment by dagargo** [comment]
I'm closing this as solved.

Feel free to reopen the issue if needed.
